aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -4,13 +4,149 @@
+// contained within has the same number of non-zero elements.
+// The batches are joined along the compressed axis. The generation of indices
+// for this matrix can be performed in a single step followed by a single step
+// conversion to restore the batch dimension.
+void dense_to_sparse_compressed_prepare_check_mask_values_batched(
+    const Layout& target_layout,
+    Tensor& values,
+    Tensor& mask,
+    const int64_t& n_batch_dim) {
+  if (n_batch_dim > 1) {
+    // For inputs with more than 1 batch dim we flatten them out.
+    // Input shape (b0, b1 ..., bn, r, c) -> (b0 * b1 * ... * bn, r ,c)
+    values = values.flatten(0, n_batch_dim - 1);
+    mask = mask.flatten(0, n_batch_dim - 1);
+  }
+
+  // For informative messaging form the name of the function
+  // to_sparse_{csr,csc,bsr,bsc}.
+  TORCH_CHECK(
+      mask.size(0) > 0,
+      "to_sparse_",
+      // We want the message to match the function name so generate the
+      // lowercase acronym for the layout
+      sparse_csr::layoutToString(target_layout, false, true),
+      ": Expected product of batch dimensions to be non-zero.");
+
+  // Compute the number of non-zero elements in the first batch, expand to full
+  // size
+  auto nse_per_batch = mask.select(0, 0).sum().expand(mask.size(0));
+  TORCH_CHECK(
+      mask.sum({-2, -1}).equal(nse_per_batch),
+      "Expect the same number of specified elements per batch.");
+
+  // We need to join batches into a matrix increasing the length of the
+  // compressed axis. This allows us to create indices for a compressed matrix
+  // and de-batch them later (two kernels). Otherwise we would have to create
+  // indices for each batch individually requiring n_batch kernels. For csr/bsr,
+  // we already have the batch dim adjacent to the compressed axis and can
+  // flatten them together. For csc/bsc, we need to transpose first.
+  // For BSR/CSR (b, r, c) -> (b*r, c)
+  // For BSC/CSC (b, c, r) -> (r, b*c)
+  AT_DISPATCH_ROW_SPARSE_COMPRESSED_LAYOUTS(
+      target_layout,
+      "dense_to_sparse_compressed",
+      [&]() {
+        values = values.flatten(0, 1);
+        mask = mask.flatten(0, 1);
+      },
+      [&]() {
+        values = values.transpose(0, 1).flatten(1, 2);
+        mask = mask.transpose(0, 1).flatten(1, 2);
+      });
+}
+
+// This function unfolds the compressed indices of a compressed sparse matrix
+// into a batched compressed sparse tensor.
+// This is analogous to an unflatten-like operation:
+// unflatten(0, {b, r}) for csr/bsr with input shape (r*b, c)
+//          (output shape (b, r, c))
+// unflatten(1, {b, c}).transpose(0,1) for csc/bsc with input shape (r, c*b)
+//          (output shape (r, b, c) unflatten, (b, r, c) unflatten + transpose)
+// This only operates on the compressed indices as the plain indices and values
+// can be manipulated as described above without special handling.
+// It is a prerequisite for the conversion that the sparsity pattern is sane for
```

Comment:
`sane` or `same`, or both? :)

Reply:
It is supposed to be Sane. The pattern must be sane, meaning each sub-matrix has the same nnz but it may have a different placement of those values.